### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by ngirard** [DoumanAsh/clipboard-master#1]
Hi,
compiling your example code gives:
```
error[E0432]: unresolved imports `clipboard_master::Master`, `clipboard_master::ClipboardHandler`, `clipboard_master::CallbackResult`
 --> src/main.rs:3:24
  |
3 | use clipboard_master::{Master, ClipboardHandler, CallbackResult};
  |                        ^^^^^^  ^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^ no `CallbackResult` in the root
  |                        |       |
  |                        |       no `ClipboardHandler` in the root
  |                        no `Master` in the root
```

**Comment by ngirard** [comment]
Hmmm... looking into your `Cargo.toml` gave me a clue: I'm on Linux and your code is Windows-only, is that so ?

If so, would you mind mentioning it in the README file ?

Also, are you planning in making your lib linux-friendly any time soon ?

Cheers

**Comment by DoumanAsh** [comment]
Yeah currently it is windows only.
Recently I restructured crate to allow multiple platforms, but I'm yet to implement it, mostly because I'm not very well familiar with Linux APIs, and on Mac there is no good way to actually track clipboard changes aside from polling clipboard. I'm not sure when I'll have time to do it myself.

I'll add mention about supported platforms, if you're familiar with Linux APIs I'd welcome PR

**Comment by DoumanAsh** [comment]
@ngirard Could you please test https://github.com/DoumanAsh/clipboard-master/commit/b54ed53e375c5f64dbfddef87de5a3fb0a24b333

I added linux support using x11_clipboard crate
Let me know if it works for you so that I could publish it

**Comment by ngirard** [comment]
Hi @DoumanAsh, thanks for your efforts !

The example code now compiles fine, but nothing is printed out when I copy something to the clipboard (expected: `Clipboard change happened!`)

**Comment by DoumanAsh** [comment]
@ngirard hmmm weird, I tried to change it to monitor only text changes https://github.com/DoumanAsh/clipboard-master/commit/b7f2308b566475d2b44de2d479b6092756222f85

Could you try again?
I'm not sure how to monitor all changes

**Comment by DoumanAsh** [comment]
@quininer hey, sorry to ping you like that, but I wonder if you could advice on how to monitor any change to clipboard on X11?
Is there some special atom that I could use? I tried to use `ATOM_ANY` but it seems I mistook it, so for now I only monitor string changes, but I'd be happy to have a way to monitor all changes

**Comment by ngirard** [comment]
@DoumanAsh it works now, although with the PRIMARY selection only. I'm afraid I'd need to monitor both PRIMARY and CLIPBOARD selections, and indeed, to monitor any change, not only string.

Sorry to bother both of you, @DoumanAsh and @quininer... clipboard management on Linux is notoriously tricky as hell (see [this rant about ICCCM](https://web.archive.org/web/20180313164303/https://lists.slug.org.au/archives/slug-chat/2001/July/msg00054.html) from Conrad Parker, the author of [xsel](http://www.vergenet.net/~conrad/software/xsel/)).

**Comment by DoumanAsh** [comment]
I'm not sure if you can monitor both selection at once.
I could probably monitor it with interval one by one, or maybe having two clipboard monitors, each with own handler

**Comment by DoumanAsh** [comment]
@ngirard looking at xsel code it seems you actually can use something like `XNextEvent` to get notification on any event, including selection (though it still limited by type of selection)

**UPD**: And looking at `x11-clipboard` source code it actually should be possible to use multiple selection as `xcb` seems basically wraps x11.

**Comment by DoumanAsh** [comment]
@ngirard I copied over code from x11-clipboard to select from both primary and clipboard selections.
But i'm not sure how to get content of clipboard for user when notification comes as `x11-clipboard` clears cliboard on load

**Comment by ngirard** [comment]
@DoumanAsh About `XNextEvent`: thanks for your finding !

Meanwhile I was looking at the source code of [CopyQ](https://github.com/hluk/CopyQ), a fully-featured, cross-platform clipboard manager.

It manages to monitor both primary and clipboard selections [here](https://github.com/hluk/CopyQ/blob/93bceed8167d58669102b28efe57b7c5be67d4b0/src/platform/x11/x11platformclipboard.cpp#L161)

I'm still unsure about how it actually manages to get the clipboard data (not a big fan of C++)

**Comment by DoumanAsh** [comment]
@ngirard I believe CopyQ uses Qt here for that, probably Qt code could be good reference

**Comment by ngirard** [comment]
Oh, really ?
Reading the code, I was having the feeling that it did it in-house, because
- `#include <QClipboard>` is only used to test against `QClipboard::Clipboard` and `QClipboard::Selection` which are [constants](https://doc.qt.io/qt-5/qclipboard.html#Mode-enum);
- `#include <QMimeData>`... well, doesn't seem to be used
- `#include <QX11Info>` is only used in `bool isSelectionIncomplete()`

**Comment by DoumanAsh** [comment]
Ah it seems I misunderstood it seeing `onChanged` override and `changed` signal, but it seems they emit this signal manually.

I guess I could took a look more.

**Comment by ngirard** [comment]
I think the data is retrieved on [this line](https://github.com/hluk/CopyQ/blob/93bceed8167d58669102b28efe57b7c5be67d4b0/src/platform/x11/x11platformclipboard.cpp#L180)
```cpp
    const auto data = ::clipboardData(clipboardData->mode);
```
I'm a bit ashamed to confess I don't understand its syntax (I mean the ``::clipboardData()``)

**Comment by DoumanAsh** [comment]
`::clipboardData()` means it is using function from within root namespace

**Comment by ngirard** [comment]
That's what I thought, but I couldn't find such `clipboardData()` function

**Comment by DoumanAsh** [comment]
I think this is the one https://github.com/hluk/CopyQ/blob/a9b14fac4150e579a2fd5a8e79e35827a934580f/src/common/common.cpp#L240

**Comment by ngirard** [comment]
Ooh, good catch !
And indeed, it relies on Qt to get the clipboard data...

**Comment by DoumanAsh** [comment]
yeah, so I'll look at it a bit later.
Currently master branch will notify you on text changes to both primary and clipboard selection, but further implementation will need more time for me as I'm on windows myself.

**Comment by ngirard** [comment]
Great, thanks !

**Comment by DoumanAsh** [comment]
@ngirard It is been a while and I completely got stuck with x11 implementation.
If you're still interested, I would appreciate if you could test current master branch to see how it works

**Comment by DoumanAsh** [comment]
I've published new version `3.1.2` with x11 fix from @cc-morning 
According to him Linux now works properly
